libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/mpp/syncKotlinAndAndroidSourceSets.kt

```diff
@@ -0,0 +1,123 @@
+ */
+
+package org.jetbrains.kotlin.gradle.plugin.mpp
+
+import com.android.build.gradle.BaseExtension
+import com.android.build.gradle.api.AndroidSourceSet
+import org.gradle.api.Project
+import org.jetbrains.kotlin.gradle.dsl.kotlinExtension
+import org.jetbrains.kotlin.gradle.internal.Kapt3GradleSubplugin
+import org.jetbrains.kotlin.gradle.plugin.*
+import org.jetbrains.kotlin.gradle.plugin.AbstractAndroidProjectHandler.Companion.kotlinSourceSetNameForAndroidSourceSet
+import org.jetbrains.kotlin.gradle.plugin.addConvention
+import java.io.File
+
+internal fun syncKotlinAndAndroidSourceSets(target: KotlinAndroidTarget) {
+    val project = target.project
+    val android = project.extensions.getByName("android") as BaseExtension
+
+    android.sourceSets.all { androidSourceSet ->
+        val kotlinSourceSetName = kotlinSourceSetNameForAndroidSourceSet(target, androidSourceSet.name)
+        val kotlinSourceSet = project.kotlinExtension.sourceSets.maybeCreate(kotlinSourceSetName)
+        createDefaultDependsOnEdges(target, kotlinSourceSet, androidSourceSet)
+        syncKotlinAndAndroidSourceDirs(target, kotlinSourceSet, androidSourceSet)
+        syncKotlinAndAndroidResources(target, kotlinSourceSet, androidSourceSet)
+        androidSourceSet.addConvention(KOTLIN_DSL_NAME, kotlinSourceSet)
+
+        ifKaptEnabled(project) {
+            Kapt3GradleSubplugin.createAptConfigurationIfNeeded(project, androidSourceSet.name)
+        }
+    }
+}
+
+private fun createDefaultDependsOnEdges(
+    target: KotlinAndroidTarget,
+    kotlinSourceSet: KotlinSourceSet,
+    androidSourceSet: AndroidSourceSet
+) {
+    val commonSourceSetName = when (androidSourceSet.name) {
+        "main" -> "commonMain"
+        "test" -> "commonTest"
+        "androidTest" -> "commonTest"
+        else -> return
+    }
+    val commonSourceSet = target.project.kotlinExtension.sourceSets.findByName(commonSourceSetName) ?: return
+    kotlinSourceSet.dependsOn(commonSourceSet)
+}
+
+private fun syncKotlinAndAndroidSourceDirs(
+    target: KotlinAndroidTarget, kotlinSourceSet: KotlinSourceSet, androidSourceSet: AndroidSourceSet
+) {
+    val disambiguationClassifier = target.disambiguationClassifier
+
+    /*
+    Mitigate ambiguity!
+    Example: disambiguationClassifier="android"
+    Source Directory "src/androidTest/kotlin"
+        -- could be claimed by kotlin {android}Test (unit test)
+        -- could be claimed by Android androidTest (instrumented test)
+
+    The Kotlin source set would win in this scenario.
+     */
+    if (disambiguationClassifier == null || !androidSourceSet.name.startsWith(disambiguationClassifier)) {
+        kotlinSourceSet.kotlin.srcDir("src/${androidSourceSet.name}/kotlin")
+    }
+
+    kotlinSourceSet.kotlin.srcDirs(*androidSourceSet.java.srcDirs.toTypedArray())
+    androidSourceSet.java.srcDirs(*kotlinSourceSet.kotlin.srcDirs.toTypedArray())
```

Comment:
I think we should avoid adding e.g `src/main/kotlin` to java source directories of AGP.

Reply:
Oh! I would be interested on why this could be harmful? 
I actually had a good intention for registering this code!

See this issue: [MPP: Instrumented Android Tests are not executed](https://youtrack.jetbrains.com/issue/KT-35016)
It seems that AGP is checking if test sources are empty before executing them on device. However, when a project does only include sources inside the kotlin source folder, then AGP does not take them into account and therefore the tests are not executed :( 

- I also experienced that test-gutters worked better when registering the source sets
- Source set colors were also correct when this is done. 

If adding the Kotln source folders here is a bad idea, then we probably should discuss a solution on AGP side?
I would be glad to help here ☺️

Reply:
The approach of registering the Kotlin source folder here is also done by the compose team: 

https://android.googlesource.com/platform/frameworks/support/+/androidx-master-dev/buildSrc/src/main/kotlin/androidx/build/AndroidXUiPlugin.kt#195

Reply:
Actually, it seems that the same is happening for non-hmpp projects!

Reply:
This is definitely an issue on the AGP side, I've just filled https://issuetracker.google.com/issues/173770818.

Reply:
This is a very good argument ☺️ 👍 
Since there are potentially other issues fixed with registering those source folders in AndroidSourceSet, we might want to have a call on that? Maybe we could offer you a better API for looking at our source sets?

Reply:
@h0tk3y What do you think?
Registering Kotlin source sets in Android Source Set "java" directory:

+ Support for .java files in /kotlin source sets (I think some would intuitively expect this)
+ Will fix tests not running when only kotlin sources present (actually pretty severe issue IMHO)
- Potential cost of scanning more files for .java sources where we do not actually expect _many_ java files to begin with

Reply:
> + Support for .java files in /kotlin source sets (I think some would intuitively expect this)
In fact, I think there could be reasons why that's not what we want in MPP. For example, if we at some point support compiling JVM/Android-shared source sets to `klib`, I'm not sure we will want Java sources mixed with Kotlin in Android-shared source sets. However, this doesn't seem like a blocker for having Kotlin and Java mixed in the source directories.

> + Will fix tests not running when only kotlin sources present (actually pretty severe issue IMHO)
I agree that it's a major issue that we should fix this sooner rather than later. Until a proper fix is implemented on the AGP side, we might even want to choose the suboptimal solution.

However, note that allowing Java sources is not something that we can freely take back afterwards. Even if we decide to do that, we will have to provide migration assistance to users who will have already added their Java sources to the Kotlin source roots.

My opinion on this is: if we are sure that the fix is going to appear on the AGP side soon enough, then not registering the Kotlin sources as Java is a bit more preferable.

Reply:
Regarding the https://issuetracker.google.com/issues/173770818, it will be fixed in next AGP alpha, and we may cherry-pick it to 4.2 branch as well, so one of the issues will be removed.

Reply:
Awesome! 🎉 
I already removed the automatic registration! 
Let's merge it without. I will try to find what other issues would have been mitigated by this change and we will find another solution afterwards ☺️